<a href="https://colab.research.google.com/github/hermiMAB/Bold-Move-Foundation/blob/main/PCA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**NOTE THAT THIS ACTIVITY WILL BE GRADED**


**In this activity you will be applying PCA techniques you have learned yesterday.**

You will be using the **Human Activity Recognition (HAR) dataset** which is a collection of data aimed at identifying human activities based on measurements from smartphone sensors. Specifically, it contains data from accelerometers and gyroscopes of smartphones carried by subjects performing activities like walking, sitting, standing, lying down, walking upstairs, and walking downstairs. This dataset is widely used for research in areas such as activity recognition, human-computer interaction, and ubiquitous computing. It provides a valuable resource for developing and testing algorithms that can automatically detect and classify human activities based on sensor data.


Run the next cell to load the dataset

In [ ]:
# Code for downloading the dataset and loading it
import pandas as pd
import numpy as np
import requests
import zipfile
import io

# URL for the dataset archive
HAR_DATASET_URL = 'https://archive.ics.uci.edu/ml/machine-learning-databases/00240/UCI%20HAR%20Dataset.zip'

# Download the dataset
print("Downloading the dataset...")
response = requests.get(HAR_DATASET_URL)
z = zipfile.ZipFile(io.BytesIO(response.content))
print("Download complete. Extracting files...")

# Extract the dataset
z.extractall()

# Define file paths for training and test sets
TRAIN_DATA_PATH = 'UCI HAR Dataset/train/X_train.txt'
TRAIN_LABELS_PATH = 'UCI HAR Dataset/train/y_train.txt'
TEST_DATA_PATH = 'UCI HAR Dataset/test/X_test.txt'
TEST_LABELS_PATH = 'UCI HAR Dataset/test/y_test.txt'

# Load the dataset into pandas DataFrames
X_train = pd.read_csv(TRAIN_DATA_PATH, delim_whitespace=True, header=None)
y_train = pd.read_csv(TRAIN_LABELS_PATH, delim_whitespace=True, header=None).squeeze()

X_test = pd.read_csv(TEST_DATA_PATH, delim_whitespace=True, header=None)
y_test = pd.read_csv(TEST_LABELS_PATH, delim_whitespace=True, header=None).squeeze()

# Feature names extraction
feature_names_path = 'UCI HAR Dataset/features.txt'
features = pd.read_csv(feature_names_path, delim_whitespace=True, header=None, usecols=[1])

# Naming the feature columns
X_train.columns = features.iloc[:, 0].values
X_test.columns = features.iloc[:, 0].values

print("Data loaded successfully.")
print("Training set size:", X_train.shape)
print("Test set size:", X_test.shape)

# Optionally, load the activity labels to make the y values more readable
activity_labels_path = 'UCI HAR Dataset/activity_labels.txt'
activity_labels = pd.read_csv(activity_labels_path, delim_whitespace=True, header=None, index_col=0)
y_train.replace(activity_labels.to_dict()[1], inplace=True)
y_test.replace(activity_labels.to_dict()[1], inplace=True)

Download complete. Extracting files...
Data loaded successfully.
Training set size: (7352, 561)
Test set size: (2947, 561)


Here is a part of the feature matrix

In [ ]:
X_train.head()

,tBodyAcc-mean()-X,tBodyAcc-mean()-Y,tBodyAcc-mean()-Z,tBodyAcc-std()-X,tBodyAcc-std()-Y,tBodyAcc-std()-Z,tBodyAcc-mad()-X,tBodyAcc-mad()-Y,tBodyAcc-mad()-Z,tBodyAcc-max()-X,...,fBodyBodyGyroJerkMag-meanFreq(),fBodyBodyGyroJerkMag-skewness(),fBodyBodyGyroJerkMag-kurtosis(),"angle(tBodyAccMean,gravity)","angle(tBodyAccJerkMean),gravityMean)","angle(tBodyGyroMean,gravityMean)","angle(tBodyGyroJerkMean,gravityMean)","angle(X,gravityMean)","angle(Y,gravityMean)","angle(Z,gravityMean)"
0,0.288585,-0.020294,-0.132905,-0.995279,-0.983111,-0.913526,-0.995112,-0.983185,-0.923527,-0.934724,...,-0.074323,-0.298676,-0.710304,-0.112754,0.030400,-0.464761,-0.018446,-0.841247,0.179941,-0.058627
1,0.278419,-0.016411,-0.123520,-0.998245,-0.975300,-0.960322,-0.998807,-0.974914,-0.957686,-0.943068,...,0.158075,-0.595051,-0.861499,0.053477,-0.007435,-0.732626,0.703511,-0.844788,0.180289,-0.054317
2,0.279653,-0.019467,-0.113462,-0.995380,-0.967187,-0.978944,-0.996520,-0.963668,-0.977469,-0.938692,...,0.414503,-0.390748,-0.760104,-0.118559,0.177899,0.100699,0.808529,-0.848933,0.180637,-0.049118
3,0.279174,-0.026201,-0.123283,-0.996091,-0.983403,-0.990675,-0.997099,-0.982750,-0.989302,-0.938692,...,0.404573,-0.117290,-0.482845,-0.036788,-0.012892,0.640011,-0.485366,-0.848649,0.181935,-0.047663
4,0.276629,-0.016570,-0.115362,-0.998139,-0.980817,-0.990482,-0.998321,-0.979672,-0.990441,-0.942469,...,0.087753,-0.351471,-0.699205,0.123320,0.122542,0.693578,-0.615971,-0.847865,0.185151,-0.043892


In [ ]:
y_train.head(-10)

,0
0,STANDING
1,STANDING
2,STANDING
3,STANDING
4,STANDING
...,...
7337,WALKING_UPSTAIRS
7338,WALKING_UPSTAIRS
7339,WALKING_UPSTAIRS
7340,WALKING_UPSTAIRS


Let's also have a look at the categories in the target labels and the number of samples in each category

In [ ]:
y_train.value_counts()

LAYING                1407
STANDING              1374
SITTING               1286
WALKING               1226
WALKING_UPSTAIRS      1073
WALKING_DOWNSTAIRS     986
Name: 0, dtype: int64

### Step 1: Apply integer encoding to target values

The target values are LAYING, STANDING, SITTING, etc. We should represent them numerically to apply automatic classification. Write code to apply such a transformation.

In [ ]:
# Put your code here and run it


### Step 2: Write code to standardize the feature matrix


In [ ]:
# Put your code here and run it


## Step 3: Visualize data on 2 dimensions

The feature vector dimension is 561 (X_train shape was: (7352, 561) ). Still you would like to visualize this data in two dimensions and you know PCA can help you. Write the code to plot train data in a 2D space.

**Hint:** You may find it practical to create a new pandas dataframe with colums for principal components and the target values and then use the Seaborn library scatter plot function:
https://seaborn.pydata.org/generated/seaborn.scatterplot.html

In [ ]:
# Put your code here and run it


## Step 4: Test the possibility of applying PCA as a preprocessing step for an automatic classification task

In this part, you should test the influence of applying PCA as a preprocessing step in an automatic classifier pipeline. Write code to compare two pipelines: one without PCA, one with PCA. Compare the performances of two pipelines trained (on train data) on the test data.

To observe the influence of the n_components in PCA, try pipelines with the following dimensions and print test results with each. n_components = 1/8 of the original dimension, similarly for scale factors: 2/8, 3/8, 4/8, 5/8 and 6/8.

To simplify the tests (and reduce testing time), **use the SVC classifier with default parameters** (without hyper-parameter optimization):

**sklearn.svm: SVC**

In [ ]:
# Put your code here and run it


## Step 6: Interpret the results

Interpret the results of your experiment with just a few sentences. Write your answer in this text cell.

